merlin/systems/dag/dictarray.py

```diff
@@ -13,44 +13,149 @@
 
         if isinstance(values, Column):
             raise TypeError("doubly nested columns")
 
         self.values = values
+        self.row_lengths = row_lengths
         self.dtype = values.dtype
 
+        if isinstance(values, np.ndarray):
+            self._device = Device.CPU
+        elif HAS_GPU and cupy and isinstance(values, cupy.ndarray):
+            self._device = Device.GPU
+        else:
+            raise TypeError("type of values argument is not supported")
+
+    def cpu(self):
+        """
+        Move the data for this column to host (CPU) memory
+
+        Returns
+        -------
+        Column
+            Same column, same data but now definitely in CPU memory
+        """
+        self.device = Device.CPU
+        return self
+
+    def gpu(self):
+        """
+        Move the data for this column to device (GPU) memory
+
+        Returns
+        -------
+        Column
+            Same column, same data but now definitely in GPU memory
+        """
+        self.device = Device.GPU
+        return self
+
+    @property
+    def device(self):
+        return self._device
+
+    @device.setter
+    def device(self, device):
+        if device == "cpu":
+            device = Device.CPU
+        elif device == "gpu":
+            device = Device.GPU
+
+        # CPU to GPU
+        if self._device == Device.CPU and device == Device.GPU:
+            self._device_move(cupy.asarray)
+        # GPU to CPU
+        elif self._device == Device.GPU and device == Device.CPU:
+            self._device_move(cupy.asnumpy)
+        # Nothing to do
+        else:
+            return self
+
+    def _device_move(self, fn):
+        if not cupy:
+            raise ValueError("Unable to move data across devices without a GPU")
+        self.values = fn(self.values)
+        if self.row_lengths:
+            self.row_lengths = fn(self.row_lengths)
+
     def __getitem__(self, index):
-        return self.values[index]
+        if self.is_list:
+            start = self._array_lib.cumsum(self.row_lengths[:index])
+            end = start + self.row_lengths[index] - 1
+            return self.values[start:end]
+        else:
+            return self.values[index]
 
     def __eq__(self, other):
-        return all(self.values == other.values) and self.dtype == other.dtype
+        values_eq = all(self.values == other.values) and self.dtype == other.dtype
+        if self.is_list:
+            return values_eq and all(self.row_lengths == other.row_lengths)
+        else:
+            return values_eq
 
     def __len__(self):
-        return len(self.values)
+        if self.is_list:
+            return len(self.row_lengths)
+        else:
+            return len(self.values)
 
     @property
     def shape(self):
-        return self.values.shape
+        if self.is_list:
+            dim = self.row_lengths[0] if self.is_ragged else None
+            return (len(self), dim)
+        else:
+            return self.values.shape
+
```

Comment:
The `is_list` and `is_ragged` properties below directly align with the schema properties, which should make validating the actual data against the schemas a bit easier than it has been up to this point.